### PR TITLE
[MI-3012]: Converted user open PRs API into GraphQL

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
https://github.com/mattermost/mattermost-plugin-github/pull/626

### CODE REVIEW

File: server/plugin/internal/graphql/client.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	httpClient := oauth2.NewClient(context.Background(), ts)
+	var client Client
+
+	if enterpriseBaseURL == "" || orgName == "" {
+		client = Client{
+			username: username,
+			client:   githubv4.NewClient(httpClient),
+		}
+	} else {
+		baseURL, _ := url.Parse(enterpriseBaseURL)
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/client.go]: why aren't we checking the error here?

File: server/plugin/internal/graphql/client.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		baseURL.Path = path.Join(baseURL.Path, "api", "graphql")
+
+		client = Client{
+			client:   githubv4.NewEnterpriseClient(baseURL.String(), httpClient),
+			username: username,
+			org:      orgName,
+		}
+	}
+
+	// For reuse
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/client.go]: i don't think there is any need for this comment

File: server/plugin/internal/graphql/client.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+	return &client
+}
+
+// executeQuery takes a query struct and sends it to GitHub GraphQL API via helper package.
+func (c *Client) executeQuery(qry interface{}, params map[string]interface{}) error {
+	if err := c.client.Query(context.Background(), qry, params); err != nil {
+		return errors.Wrap(err, "query execution error")
+	}
+	return nil
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/client.go]:
```suggestion
	}
	
	return nil
```

File: server/plugin/internal/graphql/client.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	// Set services
+	client.PullRequests = (*PullRequestService)(&common)
+
+	return &client
+}
+
+// executeQuery takes a query struct and sends it to GitHub GraphQL API via helper package.
+func (c *Client) executeQuery(qry interface{}, params map[string]interface{}) error {
+	if err := c.client.Query(context.Background(), qry, params); err != nil {
+		return errors.Wrap(err, "query execution error")
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/client.go]:
```suggestion
		return errors.Wrap(err, "error in executing query")
```

Comment (ayusht2810) [reply server/plugin/internal/graphql/client.go]: maybe?

File: server/plugin/internal/graphql/pull_request.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+// PullRequestService is responsible for handling pull request related queries
+type PullRequestService service
+
+func (p *PullRequestService) GetYourPrs() ([]*github.Issue, error) {
+	params := map[string]interface{}{
+		queryParamPRSearchQuery: githubv4.String(fmt.Sprintf("author:%s is:pr is:%s archived:false", p.client.username, githubv4.PullRequestStateOpen)),
+		queryParamPRsCursor:     (*githubv4.String)(nil),
+	}
+	if p.client.org != "" {
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/pull_request.go]:
```suggestion
	}
	
	if p.client.org != "" {
```

File: server/plugin/internal/graphql/client.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	client *Client
+}
+
+// Client encapsulates the third party package that communicates with Github GraphQL API
+type Client struct {
+	client   *githubv4.Client
+	org      string
+	username string
+
+	PullRequests *PullRequestService
```
Comment (raghavaggarwal2308) [inline server/plugin/internal/graphql/client.go:25]: why this with capital case and rest smallcase?

File: server/plugin/internal/graphql/client.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+	// For reuse
+	common := service{client: &client}
+	// Set services
+	client.PullRequests = (*PullRequestService)(&common)
+
+	return &client
+}
+
+// executeQuery takes a query struct and sends it to GitHub GraphQL API via helper package.
```
Comment (raghavaggarwal2308) [inline server/plugin/internal/graphql/client.go]:
```suggestion
// executeQuery takes a query struct and sends it to Github GraphQL API via helper package.
```
for consistancy

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: server/plugin/internal/graphql/client.go (hunk 4 above)
Comment (Kshitij-Katiyar) [reply server/plugin/internal/graphql/client.go:59]: gaand mra

File: server/plugin/internal/graphql/client.go (hunk 6 above)
Comment (Kshitij-Katiyar) [reply server/plugin/internal/graphql/client.go:25]: Here I am exporting only this variable that's why

File: server/plugin/internal/graphql/client.go [hunk 8]
```diff
@@ -33,30 +35,36 @@ func NewClient(token oauth2.Token, username, orgName, enterpriseBaseURL string)
 		client = Client{
 			username: username,
 			client:   githubv4.NewClient(httpClient),
+			api:      api,
 		}
 	} else {
-		baseURL, _ := url.Parse(enterpriseBaseURL)
+		baseURL, err := url.Parse(enterpriseBaseURL)
+		if err != nil {
+			api.LogDebug("Not able to parse url", err)
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/client.go]:
```suggestion
			api.LogDebug("Not able to parse the URL", "Error", err)
```

**Review by ayusht2810 — APPROVED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/internal/graphql/client.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+			client:   githubv4.NewClient(httpClient),
+			api:      api,
+		}
+	} else {
+		baseURL, err := url.Parse(enterpriseBaseURL)
+		if err != nil {
+			api.LogDebug("Not able to parse the URL", "Error", err)
+		}
+
+		baseURL.Path = path.Join(baseURL.Path, "api", "graphql")
```
Comment (manojmalik20) [inline server/plugin/internal/graphql/client.go:46]: Won't this line give error if there's an error in parsing the URL on line no 41?

File: server/plugin/internal/graphql/client.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+type Client struct {
+	client   *githubv4.Client
+	org      string
+	username string
+	api      plugin.API
+
+	PullRequests *PullRequestService
+}
+
+// NewClient creates and returns Client. Third party package that queries GraphQL is initialized here.
```
Comment (manojmalik20) [inline server/plugin/internal/graphql/client.go:28]:
```suggestion
// NewClient creates and returns Client. The third party package that queries GraphQL is initialized here.
```

File: server/plugin/internal/graphql/client.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+			client:   githubv4.NewEnterpriseClient(baseURL.String(), httpClient),
+			username: username,
+			org:      orgName,
+			api:      api,
+		}
+	}
+
+	common := service{client: &client}
+	// Set services
+	client.PullRequests = (*PullRequestService)(&common)
```
Comment (manojmalik20) [inline server/plugin/internal/graphql/client.go:58]: The structure here doesn't seem right. `service` struct contains `Client` inside it and the `client` struct contains `service` inside it. Try to remove this nesting cycle.

**Pull request closed without merging**